### ISSUE TITLE
Empty strings are silently replaced with 'null' now

### ISSUE THREAD

**Issue by Anton3** [jbeder/yaml-cpp#1290]
I've been upgrading from an older yaml-cpp version to 0.8.0.
What I found different is that for empty strings, behavior has changed, they are now replaced with `"null"`.

For example, given the following YAML:

```yaml
message:
```

```cpp
node["message"].as<std::string>()  //=> "null"
```

I know that such a value can be interpreted as `null`, hence `IsNull() -> true` is expected.

However, my parsing for this particular option does not allow for nulls, I always interpret this as a string. In my case `message=""` vs `message="null"` makes a difference, and I can't easily distinguish between them.

Can yaml-cpp revert to the old behavior in this case?
For an empty value, I want `Scalar() == as<std::string>() == ""`
For `null` value, I want `Scalar() == as<std::string>() == "null"`

**Comment by Anton3** [comment]
This might be okay if the new behavior is more consistent.
Just notice that some breaking changes have happened around handling nulls.

**Comment by jbeder** [comment]
You can check the type of the node. Is that sufficient?

**Comment by Anton3** [comment]
Not really, because earlier I force-interpreted the node as a string. I guess that's not possible in the recent versions of the library.

The way forward for me will be to go through my configs, replacing `<empty>` with `''` and `null` with `'null'`.

**Comment by Anton3** [comment]
Reopening because of inconsistency.

According to #590, it seems that a user may want to get the value of a Scalar (in the meaning from the [spec](https://yaml.org/spec/1.2.2/#scalar)), ignoring nulls, and that the library supports it.

But what happens in practice is, if the Scalar is a null, then its value is replaced with `"null"`, without any way to get the original value.

What should happen is, the original text representation of the null value should be kept in the `Scalar()` field. Then `as<std::string>()` should retrieve this stored `Scalar()` field.

**Comment by jbeder** [comment]
Yeah that's reasonable.

**Comment by sxrzh** [comment]
It seems that now all the Scalar is always replaced with ""(empty string) instead of "null". 

See below:
```
$ cat 1290.yaml
thisIsEmpty:
thisIsnull: null
thisIsNull: Null
thisIsNULL: NULL
thisIsStr: "null"
$ cat 1290.cpp
#include <iostream>
#include <cstdio>
#include <string>
#include <yaml-cpp/yaml.h>

int main() {
    std::string input;
    char c;
    while((c = getchar()) != EOF) input += c;

    YAML::Node node = YAML::Load(input);
    for(auto it = node.begin(); it != node.end(); ++it) {
        std::cout << it->first << ": " << it->second.as<std::string>()
                  << ", Scalar: \"" << it->second.Scalar() << "\"" << '\n';
    }
    return 0;
}
$ g++ 1290.cpp -o1290 -lyaml-cpp && ./1290 < 1290.yaml
thisIsEmpty: null, Scalar: ""
thisIsnull: null, Scalar: ""
thisIsNull: null, Scalar: ""
thisIsNULL: null, Scalar: ""
thisIsStr: null, Scalar: "null"
```

**Comment by mvanhorn** [comment]
Implemented in https://github.com/jbeder/yaml-cpp/pull/1460.

**Comment by SGSSGene** [comment]
I think there is a fundamental error in yaml-cpp. Instead of having three types Scalar, Sequence and Map it has a third type called Null (Technically, yaml-cpp also knows the type Undefined).
But Null should be a Scalar type, similar to how booleans or integers are Scalar types ([Yaml Specs](https://yaml.org/spec/1.2.2/#10211-null))

Properly fixing this breaks a lot of API.
